HW2/P3/mandelbrot.pyx

```diff
@@ -27,23 +29,105 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+    #                 mag_sq = AVX.add(AVX.mul(z_i,z_i),AVX.mul(z_r,z_r))
+    #                 if AVX.signs(AVX.greater_than(mag_sq,AVX.float_to_float8(4.0))) == all_true:
+    #                     break
+    #                 z_r_prev = z_r
+    #                 z_r = AVX.add(AVX.sub(AVX.mul(z_r_prev,z_r_prev),AVX.mul(z_i_prev,z_i_prev)),c_r)
+    #                 z_i = AVX.add(AVX.mul(AVX.float_to_float8(2.0),AVX.mul(z_r_prev,z_i_prev)),c_i)
+    #             out_counts[i, j] = iter
     with nogil:
-        for i in range(in_coords.shape[0]):
+        for i in prange(in_coords.shape[0],schedule='static',chunksize=1,num_threads=1):
             for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
-                for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
+                if j % 8 == 0:
+                    c_r = AVX.make_float8(in_real[i,j], \
+                                            in_real[i,j+1],\
+                                            in_real[i,j+2],\
+                                            in_real[i,j+3],\
+                                            in_real[i,j+4],\
+                                            in_real[i,j+5],\
+                                            in_real[i,j+6],\
+                                            in_real[i,j+7])
+                    c_i = AVX.make_float8(in_imag[i,j], \
+                                            in_imag[i,j+1],\
+                                            in_imag[i,j+2],\
+                                            in_imag[i,j+3],\
+                                            in_imag[i,j+4],\
+                                            in_imag[i,j+5],\
+                                            in_imag[i,j+6],\
+                                            in_imag[i,j+7])
+                    z_r = AVX.float_to_float8(0.0)
+                    z_i = AVX.float_to_float8(0.0)
+                    prev_sign = AVX.float_to_float8(0.0)
+                    iter_array = AVX.float_to_float8(0.0)
+                    for iter in range(max_iterations):
+                        mag_sq = AVX.add(AVX.mul(z_i,z_i),AVX.mul(z_r,z_r))
+                        # curr_sign is the float8 array of >4's
+                        curr_sign = AVX.greater_than(mag_sq,AVX.float_to_float8(4.0))
+                        t = AVX.bitwise_and(curr_sign,AVX.float_to_float8(1.0))
+                        # if you've got at least one that's not 0
+                        if AVX.signs(AVX.greater_than(AVX.sub(t,prev_sign),AVX.float_to_float8(0.0))) != 0:
+                            # take elements that have just changed to 1 and multiply then by iter and add them to iter
+                            iter_array = AVX.add(AVX.mul(AVX.sub(t,prev_sign),AVX.float_to_float8(iter)),iter_array)
```

Comment:
Code looks great and concise, but it might be helpful if you include some comments that explain what is going on (since working with an 8-tuple of values and selectively updating some of them is not that intuitive).